Source/WebCore/html/HTMLAttachmentElement.cpp

```diff
@@ -70,14 +75,122 @@ HTMLAttachmentElement::~HTMLAttachmentElement() = default;
+}
+attachment::selection {
+    background-color: blue;
+}
+attachment::-webkit-attachment-preview {
+    grid-row: 1 / 4;
+    grid-column: 1;
+    align-self: center;
+}
+attachment::-webkit-attachment-info {
+    grid-row: 1;
+    grid-column: 2;
+    font-size: x-small;
+    color: gray;
+}
+attachment::-webkit-attachment-title {
+    grid-row: 2;
+    grid-column: 2;
+    font-size: small;
+}
+attachment::-webkit-attachment-subtitle {
+    grid-row: 3;
+    grid-column: 2;
+    font-size: x-small;
+    color: rgb(82, 145, 214);
+}
+    )CSS"_s;
+
+    static MainThreadNeverDestroyed<const String> shadowStyle(StringImpl::createWithoutCopying(styleSheet.characters(), styleSheet.length()));
+
+    auto style = HTMLStyleElement::create(HTMLNames::styleTag, document(), false);
+    style->setTextContent(String { shadowStyle });
+    root.appendChild(WTFMove(style));
+
+    auto container = HTMLDivElement::create(document());
+    container->setPseudo(ShadowPseudoIds::webkitAttachmentContainer());
+    root.appendChild(container);
+
+    m_innerLegacyAttachment = adoptRef(*new HTMLAttachmentElement(HTMLNames::attachmentTag, document()));
+    m_innerLegacyAttachment->m_implementation = Implementation::LegacyUnderModern;
+    m_innerLegacyAttachment->cloneAttributesFromElement(*this);
+    m_innerLegacyAttachment->m_file = m_file;
+    m_innerLegacyAttachment->m_thumbnail = WTFMove(m_thumbnail);
+    m_innerLegacyAttachment->m_icon = WTFMove(m_icon);
+    m_innerLegacyAttachment->m_iconSize = m_iconSize;
+    m_innerLegacyAttachment->setPseudo(ShadowPseudoIds::webkitAttachmentPreview());
+    container->appendChild(*m_innerLegacyAttachment);
+
+    auto info = HTMLDivElement::create(document());
+    info->setPseudo(ShadowPseudoIds::webkitAttachmentInfo());
+    container->appendChild(info);
+
+    m_elementWithTitle = HTMLDivElement::create(document());
+    m_elementWithTitle->setPseudo(ShadowPseudoIds::webkitAttachmentTitle());
+    if (String title = attachmentTitleForDisplay(); !title.isEmpty())
+        m_elementWithTitle->setInnerText(WTFMove(title));
+    container->appendChild(*m_elementWithTitle);
+
+    m_elementWithSubtitle = HTMLDivElement::create(document());
+    m_elementWithSubtitle->setPseudo(ShadowPseudoIds::webkitAttachmentSubtitle());
+    if (String subtitle = attachmentSubtitleForDisplay(); !subtitle.isEmpty())
+        m_elementWithSubtitle->setInnerText(WTFMove(subtitle));
+    container->appendChild(*m_elementWithSubtitle);
 }
 
-RenderPtr<RenderElement> HTMLAttachmentElement::createElementRenderer(RenderStyle&& style, const RenderTreePosition&)
+RenderPtr<RenderElement> HTMLAttachmentElement::createElementRenderer(RenderStyle&& style, const RenderTreePosition& position)
 {
+    if (isModern())
+        return HTMLElement::createElementRenderer(WTFMove(style), position);
+
     return createRenderer<RenderAttachment>(*this, WTFMove(style));
 }
 
+void HTMLAttachmentElement::invalidateRender()
```

Comment:
```suggestion
void HTMLAttachmentElement::invalidateRenderer()
```

Reply:
I actually meant `invalidateRendering` (I'm not invalidating the renderer object itself). Okay with you?